radix-engine-tests/tests/intent_hash_store.rs

```diff
@@ -0,0 +1,105 @@
+use radix_engine::blueprints::intent_hash_store::EPOCHS_PER_PARTITION;
+use radix_engine::errors::RejectionError;
+use radix_engine::transaction::{ExecutionConfig, FeeReserveConfig};
+use radix_engine::types::*;
+use scrypto_unit::*;
+use transaction::builder::ManifestBuilder;
+use transaction::builder::TransactionBuilder;
+use transaction::errors::TransactionValidationError;
+use transaction::model::{
+    NotarizedTransactionV1, TransactionHeaderV1, TransactionPayload,
+    ValidatedNotarizedTransactionV1,
+};
+use transaction::signing::secp256k1::Secp256k1PrivateKey;
+use transaction::validation::{
+    NotarizedTransactionValidator, TransactionValidator, ValidationConfig,
+};
+
+#[test]
+fn test_transaction_replay_protection() {
+    let mut test_runner = TestRunner::builder().build();
+
+    let current_epoch = Epoch::of(1);
+    test_runner.set_current_epoch(current_epoch);
+    let transaction = create_notarized_transaction(TransactionParams {
+        start_epoch_inclusive: current_epoch,
+        end_epoch_exclusive: current_epoch.after(DEFAULT_MAX_EPOCH_RANGE),
+    });
+
+    // 1. Run a notarized transaction
+    let receipt = test_runner.execute_transaction(
+        get_validated(&transaction).unwrap().get_executable(),
+        FeeReserveConfig::default(),
+        ExecutionConfig::for_notarized_transaction(),
+    );
+    receipt.expect_commit_success();
+
+    // 2. Run the transaction again
+    let receipt = test_runner.execute_transaction(
+        get_validated(&transaction).unwrap().get_executable(),
+        FeeReserveConfig::default(),
+        ExecutionConfig::for_notarized_transaction(),
+    );
+    receipt.expect_specific_rejection(|e| match e {
+        RejectionError::IntentHashCommitted => true,
+        _ => false,
+    });
+
+    // 3. Update the epoch
+    let new_epoch = current_epoch.after(EPOCHS_PER_PARTITION * 190);
+    test_runner.set_current_epoch(new_epoch);
+
+    // 4. Run another transaction
```

Comment:
I'm not sure this part of the test is too useful - I'd hope building a new transaction would not be expired :) - (but perhaps changing the nonce might be a better test)?

I think we should instead ensure that the replay prevention is "complete", ie that we:
* Wait to the `end_epoch_exclusive - 1`, submit the same transaction and check that we still get `RejectionError::IntentHashCommitted` 
* And then wait another epoch to `end_epoch_exclusive`, and then submit the same transaction again here, and ensure we get a different rejection (transaction expired)

This then ensures the cover is complete.

Reply:
That way the test is asserting that the transaction can't be committed again.

Reply:
Will see if I can add a test for the described scenario.